Mathlib/Analysis/Convex/Cone/Pointed.lean

```diff
@@ -0,0 +1,189 @@
+as it allows us to use the `Module` API to work with convex cones.
+
+## TODO
+
+- Rewrite proper cones using pointed cones.
+- Construct products and/or direct sums of pointed cones.
+
+-/
+
+variable {𝕜 E F G : Type*}
+
+/-- A pointed cone is a submodule of a module with scalars restricted to being nonnegative. -/
+abbrev PointedCone (𝕜 : Type*) (E : Type*) [OrderedSemiring 𝕜] [AddCommMonoid E] [Module 𝕜 E] :=
+  Submodule { c : 𝕜 // 0 ≤ c } E
+
+namespace PointedCone
+
+-- TODO: remove `prettyPrint := false` once #6833 is merged
+local notation3 (prettyPrint := false) "𝕜≥0" => { c : 𝕜 // 0 ≤ c }
+
+section Definitions
+
+variable [OrderedSemiring 𝕜]
+variable [AddCommMonoid E] [Module 𝕜 E]
+
+instance : Coe (PointedCone 𝕜 E) (ConvexCone 𝕜 E) where
+  coe := fun S => {
+    carrier := S
+    smul_mem' := fun c hc _ hx => S.smul_mem ⟨c, le_of_lt hc⟩ hx
+    add_mem' := fun _ hx _ hy => S.add_mem hx hy }
+
+theorem coe_injective : Function.Injective ((↑) : PointedCone 𝕜 E → ConvexCone 𝕜 E) :=
+  fun _ _ => by simp
+
+@[simp]
+theorem coe_pointed (S : PointedCone 𝕜 E) : (S : ConvexCone 𝕜 E).Pointed := by
+  simp [ConvexCone.Pointed]
+
+@[ext]
+theorem ext {S T : PointedCone 𝕜 E} (h : ∀ x, x ∈ S ↔ x ∈ T) : S = T :=
+  SetLike.ext h
+
+instance instZero (S : PointedCone 𝕜 E) : Zero S :=
+  ⟨0, S.zero_mem⟩
+
+/-- The `PointedCone` constructed from a pointed `ConvexCone`. -/
+def _root_.ConvexCone.toSubmodule {S : ConvexCone 𝕜 E} (hS : S.Pointed) : Submodule 𝕜≥0 E where
+  carrier := S
+  add_mem' := fun hx hy => S.add_mem hx hy
+  zero_mem' := hS
+  smul_mem' := fun ⟨c, hc⟩ x hx => by
+    simp_rw [SetLike.mem_coe]
+    cases' eq_or_lt_of_le hc with hzero hpos
+    · unfold ConvexCone.Pointed at hS
+      convert hS
+      simp [← hzero]
+    · apply ConvexCone.smul_mem
+      convert hpos
+      exact hx
+
+@[simp]
+lemma _root_.ConvexCone.mem_toSubmodule {S : ConvexCone 𝕜 E} (hS : S.Pointed) (x : E) :
+    x ∈ S.toSubmodule hS ↔ x ∈ S :=
+  Iff.rfl
+
+@[simp]
+lemma _root_.ConvexCone.coe_toSubmodule {S : ConvexCone 𝕜 E} (hS : S.Pointed) :
+    (S.toSubmodule hS : Set E) = S :=
+  rfl
+
+end Definitions
+
+section Maps
+
+variable [OrderedSemiring 𝕜]
+variable [AddCommMonoid E] [Module 𝕜 E]
+variable [AddCommMonoid F] [Module 𝕜 F]
+variable [AddCommMonoid G] [Module 𝕜 G]
+
+/-!
+
+## Maps between pointed cones
+
+There is already a definition of maps between submodules, `Submodule.map`. In our case, these maps
+are induced from linear maps between the ambient modules that are linear over nonnegative scalars.
+Such maps are unlikely to be of any use in practice. So, we construct some API to define maps
+between pointed cones induced from linear maps between the ambient modules that are linear over
+*all* scalars.
+
+-/
+
+/-- The image of a pointed cone under a `𝕜`-linear map is a pointed cone. -/
+def map (f : E →ₗ[𝕜] F) (S : PointedCone 𝕜 E) : PointedCone 𝕜 F :=
+  let f' := LinearMap.restrictScalars 𝕜≥0 f
```

Comment:
Do you really need that `let`?

Reply:
If I just try `f` I get 

```lean 
/-
  failed to synthesize instance
    SemilinearMapClass (E →ₗ[𝕜] F) ?m.64229 E F
  -/
```

Reply:
Ok. Managed to get rid of it with a coercion.